## 38. 프로젝트 루트 확인

In [1]:
from pathlib import Path

project_root = Path.cwd()

if project_root.name == "notebooks":
    project_root = project_root.parent

data_dir = project_root / "data" / "raw"

print("프로젝트 루트:", project_root)
print("데이터 폴더:", data_dir)
print("데이터 폴더 존재:", data_dir.exists())

프로젝트 루트: c:\dev\ai-data-analysis
데이터 폴더: c:\dev\ai-data-analysis\data\raw
데이터 폴더 존재: True


## 39. 필수 파일확인

In [2]:
required_files = [
    "customers.csv",
    "products.csv",
    "orders.csv",
    "order_items.csv",

]

for file_name in required_files:
    file_path = data_dir / file_name
    print(file_name, file_path.exists(), file_path.stat().st_size if file_path.exists() else 0)


customers.csv True 5641
products.csv True 4103
orders.csv True 11692
order_items.csv True 15335


## 40. 데이터 불러오기

In [3]:

import pandas as pd

customers = pd.read_csv(data_dir / "customers.csv")
products = pd.read_csv(data_dir / "products.csv")
orders = pd.read_csv(data_dir / "orders.csv")
order_items = pd.read_csv(data_dir / "order_items.csv")


## 41. DataFrame 사전 만들기

In [4]:
datasets = {
    "customers": customers,
    "products": products,
    "orders": orders,
    "order_items": order_items,
}

print(datasets)

{'customers':      customer_id name gender  age city signup_date
0              1  김수민      F   19   광주  2024-07-17
1              2  김정호      F   32   대구  2025-11-30
2              3  이경수      F   61   성남  2024-07-10
3              4  조영호      F   55   울산  2026-05-11
4              5  이예원      F   19   부산  2024-10-11
..           ...  ...    ...  ...  ...         ...
145          146  김숙자      M   61   성남  2026-01-20
146          147  이정남      M   19   부산  2025-03-11
147          148  오도현      M   29   고양  2026-07-13
148          149  김정자      M   20   부산  2024-11-16
149          150  조미영      M   40   대전  2026-01-01

[150 rows x 6 columns], 'products':     product_id product_name category   price
0            1  전자기기 상품 001     전자기기  160000
1            2    도서 상품 002       도서   34000
2            3  전자기기 상품 003     전자기기  152000
3            4  생활용품 상품 004     생활용품   70000
4            5    식품 상품 005       식품  186000
..         ...          ...      ...     ...
95          96  생활용품 상

## 42. 크기 요약

In [5]:

shape_summary = pd.DataFrame(
    [
        {
            "dataset": name,
            "rows": df.shape[0],
            "columns": df.shape[1],
        }

        for name, df in datasets.items()

    ]

)

print(shape_summary)

       dataset  rows  columns
0    customers   150        6
1     products   100        4
2       orders   301        5
3  order_items   764        5


## 43. 한 행의 의미기록

## 데이터셋별 한 행의 의미
- customers: 고객 한 명
- products: 상품 한 개
- orders: 주문 한 건
- order_items: 주문에 포함된 상품 한 항목

## 44. 처음과 마지막 행 확인

In [6]:
for name, df in datasets.items():

    print(f"\n[{name}] 앞부분")

    display(df.head())

    print(f"[{name}] 뒷부분")

    display(df.tail())


[customers] 앞부분


,customer_id,name,gender,age,city,signup_date
0,1,김수민,F,19,광주,2024-07-17
1,2,김정호,F,32,대구,2025-11-30
2,3,이경수,F,61,성남,2024-07-10
3,4,조영호,F,55,울산,2026-05-11
4,5,이예원,F,19,부산,2024-10-11


[customers] 뒷부분


,customer_id,name,gender,age,city,signup_date
145,146,김숙자,M,61,성남,2026-01-20
146,147,이정남,M,19,부산,2025-03-11
147,148,오도현,M,29,고양,2026-07-13
148,149,김정자,M,20,부산,2024-11-16
149,150,조미영,M,40,대전,2026-01-01



[products] 앞부분


,product_id,product_name,category,price
0,1,전자기기 상품 001,전자기기,160000
1,2,도서 상품 002,도서,34000
2,3,전자기기 상품 003,전자기기,152000
3,4,생활용품 상품 004,생활용품,70000
4,5,식품 상품 005,식품,186000


[products] 뒷부분


,product_id,product_name,category,price
95,96,생활용품 상품 096,생활용품,112000
96,97,전자기기 상품 097,전자기기,154000
97,98,스포츠 상품 098,스포츠,10000
98,99,뷰티 상품 099,뷰티,200000
99,100,도서 상품 100,도서,102000



[orders] 앞부분


,order_id,customer_id,order_date,payment_method,order_status
0,1,123,2026-06-04,card,completed
1,2,77,2025-08-20,naver_pay,cancelled
2,3,138,2025-12-17,bank_transfer,cancelled
3,4,57,2026-02-27,kakao_pay,cancelled
4,5,125,2026-01-18,card,cancelled


[orders] 뒷부분


,order_id,customer_id,order_date,payment_method,order_status
296,297,22,2026-02-12,bank_transfer,completed
297,298,64,2026-01-22,naver_pay,cancelled
298,299,135,2025-09-05,bank_transfer,completed
299,300,96,2026-03-25,kakao_pay,cancelled
300,301,96,2026-03-25,kakao_pay,cancelled



[order_items] 앞부분


,order_item_id,order_id,product_id,quantity,unit_price
0,1,1,100,3,102000
1,2,1,87,5,25000
2,3,1,7,3,142000
3,4,1,9,3,193000
4,5,2,72,4,189000


[order_items] 뒷부분


,order_item_id,order_id,product_id,quantity,unit_price
759,760,297,42,4,28000
760,761,298,40,4,174000
761,762,299,8,2,189000
762,763,299,12,4,175000
763,764,300,59,5,32000


## 45. 컬럼명 확인

In [7]:
for name, df in datasets.items():
    print(name, df.columns.tolist())

customers ['customer_id', 'name', 'gender', 'age', 'city', 'signup_date']
products ['product_id', 'product_name', 'category', 'price']
orders ['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status']
order_items ['order_item_id', 'order_id', 'product_id', 'quantity', 'unit_price']


## 46. 컬럼명 공백 점검

In [8]:
for name, df in datasets.items():
    suspicious = [
        col for col in df.columns

        if col != col.strip()

    ]

    print(name, suspicious)

customers []
products []
orders []
order_items []


## 47. info() 실행

In [9]:
for name, df in datasets.items():
    print(f"\n===== {name} =====")
    df.info()


===== customers =====
<class 'pandas.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   customer_id  150 non-null    int64
 1   name         150 non-null    str  
 2   gender       150 non-null    str  
 3   age          150 non-null    int64
 4   city         150 non-null    str  
 5   signup_date  150 non-null    str  
dtypes: int64(2), str(4)
memory usage: 7.2 KB

===== products =====
<class 'pandas.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   product_id    100 non-null    int64
 1   product_name  100 non-null    str  
 2   category      100 non-null    str  
 3   price         100 non-null    int64
dtypes: int64(2), str(2)
memory usage: 3.3 KB

===== orders =====
<class 'pandas.DataFrame'>
RangeIndex: 301 entries, 0 to 300
Data columns (total 5 columns):

## 48. dtype 요약표

In [10]:
dtype_rows = []

for name, df in datasets.items():
    for column, dtype in df.dtypes.items():
        dtype_rows.append(

            {

                "dataset": name,
                "column": column,
                "dtype": str(dtype),

            }

        )

dtype_summary = pd.DataFrame(dtype_rows)

dtype_summary

,dataset,column,dtype
0,customers,customer_id,int64
1,customers,name,str
2,customers,gender,str
3,customers,age,int64
4,customers,city,str
5,customers,signup_date,str
6,products,product_id,int64
7,products,product_name,str
8,products,category,str
9,products,price,int64


## 49. 날짜형 후보 확인

In [11]:
date_columns = {
    "customers": ["signup_date"],
    "orders": ["order_date"],
}
for dataset_name, columns in date_columns.items():
    df = datasets[dataset_name]
    for column in columns:
        converted = pd.to_datetime(
            df[column],
            errors="coerce"
        )
        print(
            dataset_name,
            column,
            "원본 dtype:", df[column].dtype,
            "변환 실패:", converted.isna().sum()
        )

customers signup_date 원본 dtype: str 변환 실패: 0
orders order_date 원본 dtype: str 변환 실패: 0


## 50. 결측치 개수

In [12]:
for name, df in datasets.items():
    print(f"\n[{name}]")
    display(df.isna().sum().to_frame("missing_count"))


[customers]


,missing_count
customer_id,0
name,0
gender,0
age,0
city,0
signup_date,0



[products]


,missing_count
product_id,0
product_name,0
category,0
price,0



[orders]


,missing_count
order_id,0
customer_id,0
order_date,0
payment_method,0
order_status,0



[order_items]


,missing_count
order_item_id,0
order_id,0
product_id,0
quantity,0
unit_price,0


## 51.결측치 비율

In [13]:
missing_summaries = []
for name, df in datasets.items():
    summary = pd.DataFrame(
        {
            "missing_count": df.isna().sum(),
            "missing_rate_pct": (
                df.isna().mean() * 100
            ).round(2),

        }

    )

    summary.insert(0, "dataset", name)

    summary.insert(1, "column", summary.index)

    missing_summaries.append(summary.reset_index(drop=True))

missing_summary = pd.concat(
    missing_summaries,
    ignore_index=True
)
missing_summary

,dataset,column,missing_count,missing_rate_pct
0,customers,customer_id,0,0.0
1,customers,name,0,0.0
2,customers,gender,0,0.0
3,customers,age,0,0.0
4,customers,city,0,0.0
5,customers,signup_date,0,0.0
6,products,product_id,0,0.0
7,products,product_name,0,0.0
8,products,category,0,0.0
9,products,price,0,0.0


## 52.전체 행 중복

In [14]:
for name, df in datasets.items():
    print(name, "전체 중복:", df.duplicated().sum())

customers 전체 중복: 0
products 전체 중복: 0
orders 전체 중복: 0
order_items 전체 중복: 0


## 53. 주요 ID점검

In [15]:
primary_keys = {
    "customers": "customer_id",
    "products": "product_id",
    "orders": "order_id",
    "order_items": "order_item_id",
}

id_check_rows = []
for name, key in primary_keys.items():
    df = datasets[name]
    id_check_rows.append(
        {
            "dataset": name,
            "key": key,
            "missing": df[key].isna().sum(),
            "duplicates": df[key].duplicated().sum(),
            "unique": df[key].nunique(),
            "rows": len(df),
        }
    )
id_check_summary = pd.DataFrame(id_check_rows)
id_check_summary

,dataset,key,missing,duplicates,unique,rows
0,customers,customer_id,0,0,150,150
1,products,product_id,0,0,100,100
2,orders,order_id,0,0,301,301
3,order_items,order_item_id,0,0,764,764


## 54. 데이터 품질 발견 기록

## 데이터 품질 발견 기록

| 데이터셋 | 컬럼 | 문제 종류 | 건수 | 해석 | 처리 여부 |
|---|---|---|---:|---|---|
| customers | signup_date | 날짜 변환 실패 | 0 | 가입일자 이상 없음 | N/A |
| orders | order_date | 날짜 변환 실패 | 0 | 가입주문일자 이상 없음 | N/A |
| 해당 데이터셋 | 해당 컬럼 | 결측치 | 0 | 결측 없음 | N/A |
| 해당 데이터셋 | 전체 행 | 전체 중복 | 0 | 중복 없음 | N/A |
| 해당 데이터셋 | 주요 ID | ID 중복 | 0 | ID 중복 없음 | N/A |


## 55. 숫자형 기본 통계


In [16]:
print(customers[["age"]].describe())
print(products[["price"]].describe())
print(order_items[["quantity", "unit_price"]].describe())

              age
count  150.000000
mean    42.086667
std     15.613166
min     19.000000
25%     29.000000
50%     40.000000
75%     57.000000
max     69.000000
               price
count     100.000000
mean   110040.000000
std     56433.910574
min      5000.000000
25%     65750.000000
50%    112000.000000
75%    161000.000000
max    200000.000000
         quantity     unit_price
count  764.000000     764.000000
mean     3.053665  108561.518325
std      1.410873   56996.770604
min      1.000000    5000.000000
25%      2.000000   62000.000000
50%      3.000000  111000.000000
75%      4.000000  161250.000000
max      5.000000  200000.000000


In [17]:
#56. 범주형 값 확인
category_columns = {
    "customers": ["gender", "city"],
    "products": ["category"],
    "orders": ["payment_method", "order_status"],

}

for dataset_name, columns in category_columns.items():
    df = datasets[dataset_name]
    for column in columns:
        print(f"\n[{dataset_name}.{column}]")
        display(
            df[column]
            .value_counts(dropna=False)
            .to_frame("count")
        )


[customers.gender]


,count
gender,
F,84
M,66



[customers.city]


,count
city,
성남,21
광주,17
부산,16
대구,15
서울,15
울산,14
인천,14
대전,14
수원,13



[products.category]


,count
category,
스포츠,19
전자기기,17
생활용품,16
뷰티,16
도서,14
패션,11
식품,7



[orders.payment_method]


,count
payment_method,
kakao_pay,80
naver_pay,77
bank_transfer,74
card,70



[orders.order_status]


,count
order_status,
completed,184
cancelled,65
refunded,52


In [18]:
#57. 날짜 범위 확인
signup_dates = pd.to_datetime(
    customers["signup_date"],
    errors="coerce"
)
order_dates = pd.to_datetime(
    orders["order_date"],
    errors="coerce"

)

print("가입일 변환 실패:", signup_dates.isna().sum())

print("가입일 범위:", signup_dates.min(), "~", signup_dates.max())

print("주문일 변환 실패:", order_dates.isna().sum())

print("주문일 범위:", order_dates.min(), "~", order_dates.max())

가입일 변환 실패: 0
가입일 범위: 2023-08-11 00:00:00 ~ 2026-07-25 00:00:00
주문일 변환 실패: 0
주문일 범위: 2025-08-06 00:00:00 ~ 2026-08-05 00:00:00


## 58. 날짜 관계 점검 준비
주문일과 고객 가입일을 직접 비교하려면 두 데이터를 연결해야 합니다.

이번 장에서는 구조와 키가 연결 가능한지 확인하고, 실제 병합과 분석은 이후 장에서 본격적으로 진행합니다.

In [19]:
# 59. 참조하지 못한 고객 ID

unknown_customer_ids = sorted(
    set(orders["customer_id"].dropna())
    - set(customers["customer_id"].dropna())

)

print("존재하지 않는 customer_id:", unknown_customer_ids)

존재하지 않는 customer_id: []


## 60. 참조하지 못한 주문 ID

In [20]:
unknown_order_ids = sorted(
    set(order_items["order_id"].dropna())
    - set(orders["order_id"].dropna())
)
print("존재하지 않는 order_id:", unknown_order_ids)

존재하지 않는 order_id: []


## 61. 참조하지 못한 상품 ID

In [21]:
unknown_product_ids = sorted(
    set(order_items["product_id"].dropna())
    - set(products["product_id"].dropna())
)
print("존재하지 않는 product_id:", unknown_product_ids)

존재하지 않는 product_id: []


## 62. 반복 점검 함수

In [22]:
def inspect_dataframe(
    name: str,
    df: pd.DataFrame,
    primary_key: str | None = None,
) -> dict:
    result = {
        "dataset": name,
        "rows": df.shape[0],
        "columns": df.shape[1],
        "column_names": df.columns.tolist(),
        "dtypes": {
            column: str(dtype)
            for column, dtype in df.dtypes.items()
        },
        "missing_total": int(df.isna().sum().sum()),
        "duplicate_rows": int(df.duplicated().sum()),
    }
    if primary_key is not None:
        result["primary_key"] = primary_key
        result["key_missing"] = int(
            df[primary_key].isna().sum()
        )
        result["key_duplicates"] = int(
            df[primary_key].duplicated().sum()
        )
        result["key_unique"] = int(
            df[primary_key].nunique()
        )
    return result

In [23]:
inspection_results = []

for name, df in datasets.items():
    inspection_results.append(
        inspect_dataframe(
            name,
            df,
            primary_keys[name],
        )
    )

inspection_results

[{'dataset': 'customers',
  'rows': 150,
  'columns': 6,
  'column_names': ['customer_id',
   'name',
   'gender',
   'age',
   'city',
   'signup_date'],
  'dtypes': {'customer_id': 'int64',
   'name': 'str',
   'gender': 'str',
   'age': 'int64',
   'city': 'str',
   'signup_date': 'str'},
  'missing_total': 0,
  'duplicate_rows': 0,
  'primary_key': 'customer_id',
  'key_missing': 0,
  'key_duplicates': 0,
  'key_unique': 150},
 {'dataset': 'products',
  'rows': 100,
  'columns': 4,
  'column_names': ['product_id', 'product_name', 'category', 'price'],
  'dtypes': {'product_id': 'int64',
   'product_name': 'str',
   'category': 'str',
   'price': 'int64'},
  'missing_total': 0,
  'duplicate_rows': 0,
  'primary_key': 'product_id',
  'key_missing': 0,
  'key_duplicates': 0,
  'key_unique': 100},
 {'dataset': 'orders',
  'rows': 301,
  'columns': 5,
  'column_names': ['order_id',
   'customer_id',
   'order_date',
   'payment_method',
   'order_status'],
  'dtypes': {'order_id': 'int6

In [24]:
# 63. LLM에 전달할 안전한 구조 요약
safe_schema_summary = {
    "datasets": inspection_results,
    "relationships": {
        "orders.customer_id": {
            "references": "customers.customer_id",
            "unknown_count": len(unknown_customer_ids),
        },
        "order_items.order_id": {
            "references": "orders.order_id",
            "unknown_count": len(unknown_order_ids),
        },
        "order_items.product_id": {
            "references": "products.product_id",
            "unknown_count": len(unknown_product_ids),
        },
    },
}
print(safe_schema_summary)

{'datasets': [{'dataset': 'customers', 'rows': 150, 'columns': 6, 'column_names': ['customer_id', 'name', 'gender', 'age', 'city', 'signup_date'], 'dtypes': {'customer_id': 'int64', 'name': 'str', 'gender': 'str', 'age': 'int64', 'city': 'str', 'signup_date': 'str'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'customer_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 150}, {'dataset': 'products', 'rows': 100, 'columns': 4, 'column_names': ['product_id', 'product_name', 'category', 'price'], 'dtypes': {'product_id': 'int64', 'product_name': 'str', 'category': 'str', 'price': 'int64'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'product_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 100}, {'dataset': 'orders', 'rows': 301, 'columns': 5, 'column_names': ['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status'], 'dtypes': {'order_id': 'int64', 'customer_id': 'int64', 'order_date': 'str', 'payment_method': 'str', 'order_status'

## 64. LLM 구조 점검  프롬프트
# 온라인 쇼핑몰 데이터 구조 및 품질 점검 결과

## 1. 데이터셋별 행의 의미

| 데이터셋          | 행의 의미                | 주요 식별자          |
| ------------- | -------------------- | --------------- |
| `customers`   | 고객 1명을 의미            | `customer_id`   |
| `products`    | 상품 1개를 의미            | `product_id`    |
| `orders`      | 주문 1건을 의미할 가능성이 높음   | `order_id`      |
| `order_items` | 주문에 포함된 상품 항목 1개를 의미 | `order_item_id` |

* `orders`는 **주문 단위**, `order_items`는 **주문에 포함된 상품 항목 단위**로 데이터의 집계 단위가 다름.
* `orders` 301건, `order_items` 764건으로 주문당 평균 약 **2.54개의 상품 항목**이 존재함.
* 따라서 향후 매출 및 판매량 분석 시 어떤 데이터셋을 기준으로 집계하는지 구분할 필요가 있음.

---

## 2. 기본키 및 외래키 구조

### 기본키(Primary Key)

| 데이터셋          | 기본키             | 점검 결과         |
| ------------- | --------------- | ------------- |
| `customers`   | `customer_id`   | 결측 0건 / 중복 0건 |
| `products`    | `product_id`    | 결측 0건 / 중복 0건 |
| `orders`      | `order_id`      | 결측 0건 / 중복 0건 |
| `order_items` | `order_item_id` | 결측 0건 / 중복 0건 |

각 데이터셋의 기본키는 행 수와 고유값 수가 동일하여 현재 구조상 기본키 문제는 발견되지 않음.

### 외래키(Foreign Key)

| 외래키                      | 참조 대상                   | 참조 오류 |
| ------------------------ | ----------------------- | ----: |
| `orders.customer_id`     | `customers.customer_id` |    0건 |
| `order_items.order_id`   | `orders.order_id`       |    0건 |
| `order_items.product_id` | `products.product_id`   |    0건 |

현재 확인된 범위에서는 외래키의 미등록 참조가 없어 **참조 무결성이 정상적으로 유지되고 있음**.

> `orders.customer_id`는 외래키이므로 여러 번 나타날 수 있으며, 동일 고객의 반복 주문은 정상적인 데이터일 수 있음.

---

## 3. 결측치 및 중복 데이터 처리 시 주의사항

현재 네 데이터셋 모두 **결측치 0건**, **전체 행 중복 0건**으로 확인됨.

따라서 현재 단계에서는 삭제할 결측치나 중복 데이터가 없음.

향후 결측치나 중복이 발견되더라도 즉시 삭제하지 않고 발생 원인을 먼저 확인해야 함.

### 중복으로 보이지만 정상일 수 있는 사례

* `orders.customer_id`의 반복
  → 동일 고객의 재구매일 가능성이 있음.
* `order_items`에서 동일한 `order_id + product_id`가 반복
  → 실제 주문 구조에 따라 정상적인 기록일 가능성이 있으므로 원인 확인 필요.
* `customers.name`의 반복
  → 동명이인일 가능성이 있으므로 이름만으로 중복 고객으로 판단하면 안 됨.

### 결측치 처리 시 주의

* 결측값이 실제 데이터 부재를 의미하는지 확인할 필요가 있음.
* 특정 `order_status`나 거래 상황 때문에 발생한 결측값일 가능성도 있으므로 원인 확인 후 처리해야 함.

**원칙: 결측치나 중복이 발견되더라도 원인을 확인하기 전에는 임의로 삭제하지 않는다.**

---

## 4. 추가 구조 및 품질 점검 우선순위

### 🔴 1순위 — `orders` 행 수 변경 여부 확인

현재 `orders`는 **301행**으로 확인됨.

이전 점검 결과와 행 수가 달라졌다면 다음 가능성을 확인할 필요가 있음.

* 실제 데이터가 추가되었는지
* 다른 파일을 불러온 것인지
* 데이터 필터 조건이 달라졌는지
* 이전 점검 결과가 현재 데이터와 섞였는지

현재 요약에서는 `orders.customer_id`의 미등록 참조가 **0건**으로 확인되므로, 과거 점검 결과와 현재 결과가 다르다면 최신 데이터를 기준으로 다시 확인해야 함.

**확인 코드의 목적:** 현재 사용 중인 데이터의 행 수와 ID 상태가 점검 요약과 일치하는지 확인한다.

---

### 🔴 2순위 — `unit_price`와 `products.price` 비교

가격 정보가 다음 두 데이터셋에 존재함.

* `products.price`
* `order_items.unit_price`

두 값이 다른 경우 할인 가격, 가격 변경, 또는 데이터 오류 등 여러 가능성이 있으므로 원인을 확인할 필요가 있음.

**확인 코드의 목적:** `product_id`를 기준으로 상품 가격을 연결한 후 `unit_price`와 `price`가 다른 주문 상품 항목의 수를 확인한다.

---

### 🟠 3순위 — 고객 가입일과 주문일 비교

다음 날짜 정보를 비교할 필요가 있음.

* `customers.signup_date`
* `orders.order_date`

고객 가입일보다 주문일이 빠른 데이터가 존재한다면 데이터 입력 또는 날짜 처리 과정에서 문제가 발생했을 가능성이 있음.

현재 두 컬럼의 자료형이 `str`이므로 날짜형으로 변환한 후 비교해야 함.

**확인 코드의 목적:** 가입일과 주문일을 날짜형으로 변환하여 주문일이 가입일보다 빠른 사례가 있는지 확인한다.

---

### 🟠 4순위 — 주문 상태별 거래 금액 확인

`orders`에는 `order_status`가 존재하므로 주문 상태에 따라 거래 금액을 어떻게 처리할지 확인할 필요가 있음.

다만 현재 구조 요약에는 실제 `order_status` 값이 제시되지 않았으므로 특정 상태가 존재한다고 단정할 수 없음.

**확인 코드의 목적:** 주문 상태별 주문 건수와 주문 상품 금액을 집계하여 상태에 따른 거래 규모를 확인한다.

---

### 🟡 5순위 — 상품 항목이 없는 주문 확인

`orders`에 존재하는 `order_id`가 `order_items`에 한 번도 나타나지 않는 경우가 있는지 확인할 필요가 있음.

외래키 참조 오류가 0건이라는 것과 **모든 주문에 상품 항목이 존재한다는 것은 서로 다른 점검**이므로 별도로 확인해야 함.

**확인 코드의 목적:** `orders`에는 존재하지만 `order_items`에는 존재하지 않는 `order_id`를 확인한다.

---

### 🟡 6순위 — 주문 상품 조합 중복 확인

`order_items`의 `order_id + product_id` 조합이 반복되는지 확인할 필요가 있음.

단, 반복 자체를 오류로 판단하지 않고 실제 주문 구조상 정상인지 추가 확인해야 함.

**확인 코드의 목적:** `order_id`와 `product_id`의 조합이 반복되는 행의 수를 확인한다.

---

### 🟢 7순위 — 미주문 고객 및 미판매 상품 확인

모든 고객이 주문한 것은 아닐 수 있으며, 모든 상품이 판매된 것도 아닐 수 있음.

따라서 미주문 고객이나 미판매 상품은 **데이터 오류가 아니라 실제 데이터 특성일 가능성**이 있음.

**확인 코드의 목적:** 주문 및 판매 기록에 한 번도 등장하지 않는 고객 ID와 상품 ID의 수를 확인한다.

---

## 5. 현재 점검 결과 요약

| 점검 항목                          |   결과 | 판단              |
| ------------------------------ | ---: | --------------- |
| 전체 결측치                         |   0건 | 이상 없음           |
| 전체 행 중복                        |   0건 | 이상 없음           |
| 기본키 결측                         |   0건 | 이상 없음           |
| 기본키 중복                         |   0건 | 이상 없음           |
| `orders → customers` 참조 오류     |   0건 | 이상 없음           |
| `order_items → orders` 참조 오류   |   0건 | 이상 없음           |
| `order_items → products` 참조 오류 |   0건 | 이상 없음           |
| `orders` 행 수                   | 301건 | 이전 결과와 차이 확인 필요 |
| 상품 가격 불일치                      |  미확인 | 추가 점검 필요        |
| 가입일/주문일 역전                     |  미확인 | 추가 점검 필요        |
| 주문 상태별 거래 금액                   |  미확인 | 추가 점검 필요        |
| 상품 항목이 없는 주문                   |  미확인 | 추가 점검 필요        |

---

## 6. 종합 판단

현재 구조 점검 결과, **기본키·결측치·전체 행 중복·외래키 참조 무결성에서는 명확한 문제가 발견되지 않음.**

다만 `orders`의 행 수가 이전 점검 결과와 달라졌으므로 **현재 사용 중인 데이터가 최신 데이터인지 먼저 확인할 필요가 있음.**

이후 다음 순서로 추가 점검을 진행하는 것이 적절함.

**① 데이터 기준점 확인 → ② 상품 가격 비교 → ③ 가입일/주문일 비교 → ④ 주문 상태별 금액 확인 → ⑤ 상품 항목이 없는 주문 확인 → ⑥ 주문 상품 조합 중복 확인 → ⑦ 미주문 고객 및 미판매 상품 확인**

또한 현재 확인되지 않은 내용을 데이터 오류로 단정하지 않고, **'현재 확인된 사실'과 '추가 검증이 필요한 항목'을 구분하여 기록하는 것이 중요함.**
